### PR TITLE
Add push notifications

### PR DESCRIPTION
**Pull request by Charlesworth**
Add push notification feature to the Go library.

How:
Added Client.Notify method and the PushNotification type.

Ticket: 
https://pusher2.atlassian.net/browse/PROD-1426

### CODE REVIEW

**Comment by Charlesworth** [comment]
intergration tests seem to be failing for Go1.5 and lower, to do with the TestBrokenUrl test

**Comment by vivangkumar** [comment]
Push Notifications are already included in this branch. I've just not spent enough time to clean it up and merge. https://github.com/pusher/pusher-http-go/pull/24/

**Comment by Charlesworth** [comment]
fair enough, we can close this PR then and you'll merge the 24?

**Comment by vivangkumar** [comment]
We could just use this instead of the other one. The other branch requires more attention and need to be cleaned up quite a bit and I'm afraid I don't have the time currently.

Although, if there are clients using that branch for push notifications, this will be a breaking change and we should maintain that interface. Probably worth extracting push notifications related code from that branch and merging it here. WDYT?

**Comment by Charlesworth** [comment]
hmmm, not sure, will the 2.0 branch change any other current interfaces? If 2.0 will be different and breaking so it doesn't really make much difference, though if it will implement the same interfaces then we can change this pull

**Comment by Charlesworth** [comment]
I've had a look and compared the two branches and there are a few differences in implementation.

 new-lib branch implements:
`Notify(interests []string, notification *Notification) (*NotifyResponse, error)`

While this branch implements:
`Notify(pushNotification PushNotification) error`

As you can see the new-lib splits the interests from the notification while this PR includes the interests in the notification type. 

The new-lib [also provides concrete types for the APNS/GCM](https://github.com/pusher/pusher-http-go/blob/new-lib/notifications.go) (not FCM, it might be too old), while [this PR accepts an interface{} for APNS/GCM/FCM. ](https://github.com/pusher/pusher-http-go/blob/add_push_notifications/push_notifications.go#L14) Both have their advantages so might be worth a convo about how we want to implement this going forward @vivangkumar?

**Comment by Charlesworth** [comment]
Proposed changes:
- Use the `Notify(interests []string, notification *Notification) (*NotifyResponse, error)` method
- Keep using interface{} for inputting the APNS/GCM/FCM notifications

**Comment by Charlesworth** [comment]
One big change worth mentioning, responses no longer error on any 2** status codes. The tests all pass but this may change the behavior of other responses if any other 2** codes are used elsewhere

File: client.go [hunk 1]
```diff
@@ -439,3 +441,45 @@ func (c *Client) Webhook(header http.Header, body []byte) (*Webhook, error) {
...
+		&pushNotification,
+	}
+
+	err := pNRequest.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	if c.PushNotificationHost == "" {
+		return nil, errors.New("no Client.PushNotificationHost provided")
```
Comment (vivangkumar) [inline client.go]: I'd just word this as "PushNotificationHost not provided"

File: client.go [hunk 2]
```diff
@@ -439,3 +441,45 @@ func (c *Client) Webhook(header http.Header, body []byte) (*Webhook, error) {
...
+	}
+
+	requestBody, err := json.Marshal(pNRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("/%s/%s/apps/%s/notifications", PushNotifAPIPrefixDefault, PushNotifAPIVersionDefault, c.AppId)
+
+	url, err := createRequestURL("POST", c.PushNotificationHost, path, c.Key, c.Secret, authTimestamp(), c.Secure, requestBody, nil, c.Cluster)
```
Comment (vivangkumar) [inline client.go]: use the http.POST constant?

File: push_notifications.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package pusher
+
+import "errors"
+
+const (
+	WebhookLvlInfo             = "INFO"
+	WebhookLvlDebug            = "DEBUG"
+	PushNotiHostDefault        = "nativepush-cluster1.pusher.com"
```
Comment (vivangkumar) [inline push_notifications.go]: `PushNoti` -> `PushNotif`

File: push_notifications.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+type notificationRequest struct {
+	Interests []string `json:"interests"`
+	*PushNotification
+}
+
+// validate checks the notificationRequest has 0<Interests<11 and has a
+// APNS, GCM or FCM payload
+func (pN *notificationRequest) validate() error {
```
Comment (vivangkumar) [inline push_notifications.go]: I wouldn just use `pn` instead of `pN`. It is more idiomatic.

File: client.go (hunk 2 above)
Comment (Charlesworth) [inline client.go]: This sucks but using http.MethodPost actually breaks the librabry for go versions < 1.6, they added the const methods then!

**Pull request merged**